enable fluent assertions on size of list (#301, #589) and map

PascalSchumacher opened the pull request:
I adopted the navigable assertion idea of assertj-db. 

It's different from assertj's current api, but having the option to use the following assertion chain:

```java
List<Integer> ints = Arrays.asList(1, 2, 3);
assertThat(ints).size().isGreaterThan(0).isLessThanOrEqualTo(3).returnToList().contains(1).doesNotContain(4);
```

is interesting I think. (This is a fixed version of #630. I had to create a new pull request as I could not reopen the old one.)

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractSizeAssert.java:
```diff
@@ -0,0 +1,24 @@
+    super(actual, selfType);
+  }
+
+  public abstract AbstractListAssert<S, A, T> returnToList();
```
would it be possible to simply return `S` ?

PascalSchumacher replied:
Seems possible. :)

joel-costigliola commented:
I think this should not be limited to `List`, we should provide the same feature for `Iterable` (although it does not provide a `size()` it is possible to get it with `IterableUtil.sizeOf`.

This would involve changing `returnToList` to `returnToIterable`.

PascalSchumacher commented:
I agree that this would also be nice for `Iterable`s, but when I do the change I get:

>java.lang.ClassCastException: org.assertj.core.api.ListAssert cannot be cast to org.assertj.core.api.IterableAssert
	at org.assertj.core.api.IterableSizeAssert.returnToIterable(IterableSizeAssert.java:27)
	at org.assertj.core.api.list.ListAssert_size_Test.should_be_able_to_use_integer_assertions_on_size_of_list(ListAssert_size_Test.java:34)

Any ideas how to make this possible are greatly appreciated.

PascalSchumacher commented:
But even if the exception can be prevented, there is a major disadvantage to this approach:

You can only use `IterableAssert` (not `ListAssert`) on lists.

I think a way to solve this would be to make both `AbstractIterableAssert` and `AbstractListAssert` implement `#size`, but that would mean `AbstractListAssert` can not extend `AbstractIterableAssert`.

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractIterableAssert.java:
```diff
@@ -1515,4 +1517,10 @@ public S withFailMessage(String newErrorMessage, Object... args) {
   }
+  
+  @SuppressWarnings({ "rawtypes", "unchecked" })
+  public IterableSizeAssert<T, Integer> size() {
```
try to declare returning an `AbstractIterableAssert`, I don't see a need to return an `IterableSizeAssert` (but keep building an `IterableSizeAssert`) and see if it solved the cast pb...

joel-costigliola commented on src/main/java/org/assertj/core/api/IterableSizeAssert.java:
```diff
@@ -0,0 +1,29 @@
+  }
+
+  @Override
+  public IterableAssert<T> returnToIterable() {
```
as per my other comment, return `AbstractIterableAssert`

joel-costigliola commented on src/main/java/org/assertj/core/api/ListSizeAssert.java:
```diff
@@ -0,0 +1,30 @@
+  }
+  
+  @Override
+  public ListAssert<T> returnToList() {
```
return `AbstractListAssert`

PascalSchumacher commented:
@joel-costigliola Thanks a lot for the help! Now it works. :)

I guess there is no easy way around this limitation:

>You can only use IterableAssert (not ListAssert) on lists.

(after calling `size`)?

PascalSchumacher commented:
Updated according to @joel-costigliola great suggestions and added the feature for maps as well (as suggested by @jstrachan in https://github.com/joel-costigliola/assertj-assertions-generator/issues/67).

joel-costigliola commented:
Integrated (just added javadoc).
Thanks Pascal.

PascalSchumacher commented:
Thanks for merging and for adding the javadoc. :)
